src/activation/modOptionsHelpers.ts

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (C) 2024 PixieBrix, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import { isDatabaseField } from "@/components/fields/schemaFields/fieldTypeCheckers";
+import { isEmpty } from "lodash";
+import { isUUID } from "@/types/helpers";
+import type { ModDefinition } from "@/types/modDefinitionTypes";
+import type { UUID } from "@/types/stringTypes";
+import type { OptionsArgs } from "@/types/runtimeTypes";
+
+/**
+ * Create databases for any mod options database fields where the schema format is "preview", and the field value
+ * is a string to use as the database name.
+ *
+ * Modifies optionsArgs in place.
+ *
+ * @param modDefinition the mod definition
+ * @param optionsArgs the current mod options
+ * @param databaseFactory function to create a database
+ * @throws {Error} if any of the databases cannot be created
+ */
+export async function autoCreateDatabaseOptionsArgsInPlace(
+  modDefinition: ModDefinition,
+  optionsArgs: OptionsArgs,
+  databaseFactory: (args: { name: string }) => Promise<UUID>,
+): Promise<OptionsArgs> {
+  const optionsNames = Object.entries(
+    modDefinition.options?.schema?.properties ?? {},
+  )
+    .filter(
+      ([name, fieldSchema]) =>
+        typeof fieldSchema !== "boolean" &&
+        isDatabaseField(fieldSchema) &&
+        fieldSchema.format === "preview" &&
+        typeof optionsArgs[name] === "string" &&
+        !isEmpty(optionsArgs[name]) &&
+        // If the value is a UUID, then it's a database ID for an existing database
+        !isUUID(optionsArgs[name] as string),
+    )
+    .map(([name]) => name);
+
+  await Promise.all(
+    optionsNames.map(async (name) => {
+      optionsArgs[name] = await databaseFactory({ name });
```

Comment:
I think this might break mod auto-activation unless you do this:
```suggestion
      optionsArgs[name] = await databaseFactory({ name: optionsArgs[name] });
```

Reply:
Basically, it looks at mod activation option field schemas that have the DB field type `$ref`, and `format: "preview"`, and then it looks in the current activation options for those option field names. It takes the current value in the options as the DB name, creates the DB, then swaps the activation option value with the new DB id, and activates the mod.

A mod activated manually, with an existing DB, would just have that DB's id in the same activation option field.